src/distributed-transaction/percolator.md

```diff
@@ -0,0 +1,156 @@
+|-----|----------|----------|-----------|
+| Bob | 8:<br/>7:$3<br/>6:<br>5:$10 | 8:<br/>7:<br/>6:<br/>5: | 8:data@7<br/>7:<br/>6:data@5<br/>5: |
+| Joe | 7:$9<br/>6:<br/>5:$2 | 7:primary@Bob.bal<br/>6:<br/>5: | 7:<br/>6:data@5<br/>5: |
+
+3. Commit all secondary locks to complete the writing process.
+
+| key | bal:data | bal:lock | bal:write |
+|-----|----------|----------|-----------|
+| Bob | 8:<br/>7:$3<br/>6:<br>5:$10 | 8:<br/>7:<br/>6:<br/>5: | 8:data@7<br/>7:<br/>6:data@5<br/>5: |
+| Joe | 8:<br/>7:$9<br/>6:<br/>5:$2 | 8:<br/>7:<br/>6:<br/>5: | 8:data@7<br/>7:<br/>6:data@5<br/>5: |
+
+### Reading
+
+Reading from Percolator also requires a timestamp. The procedure to perform a read operation is as follows:
+
+1. Get a timestamp, `ts`.
+2. Check if the row we are going to read is locked with a timestamp in the range `[0, ts]`;
+    * If there is a timestamp in range `[0, ts]`, it means the row is locked by an earlier-started transaction. Then we are not sure whether that transaction will be committed before or after `ts`. 
+    * If there is not such a timestamp, the read can continue.
+3. Get the latest record in the row's `write` column whose `commit_ts` is in range `[0, ts]`. The record contains the `start_ts` of the transaction when it was committed.
+4. Get the row's value in the `data` column whose timestamp is exactly `start_ts`. Then the value is what we want.
+
+For example, consider this table again:
+
+| key | bal:data | bal:lock | bal:write |
+|-----|----------|----------|-----------|
+| Bob | 8:<br/>7:$3<br/>6:<br>5:$10 | 8:<br/>7:<br/>6:<br/>5: |8:data@7<br/>7:<br/>6:data@5<br/>5: |
+| Joe | 7:$9<br/>6:<br/>5:$2 | 7:primary@Bob.bal<br/>6:<br/>5: | 7:<br/>6:data@5<br/>5: |
+
+Let's read Bob's balance.
+1. Get a timestamp. Assume it's `9`.
+2. Check the lock of the row. The row of Bob is not locked, so we continue.
+3. Get the latest record in the `write` column committed before `9`. We get a record with `commit_ts` equals to `8`, and`start_ts` `7`, which means, its corresponding data is at timestamp `7` in the `data` column.
+4. Get the value in the `data` column with timestamp `7`. `$3` is the result to the read.
+
+This algorithm provides us both lock-free reading and the ability to perform historical read. In the above example, if we specify that we want to read at time point `7`, then we will see the write record at timestamp `6`, giving us the result `$10` at timestamp `5`.
+
+### Handling Conflicts
+
+Conflicts are identified by checking the `lock` column. A row can have many versions of data, but it can have at most one lock at any time.
+
+When we are performing a write operation, we try to lock every affected row in the `Prewrite` phase. If we failed to lock some of these rows, the whole transaction will be rolled back. As an optimistic lock algorithm, sometimes it causes a shortage of Percolator: there may be performance regression in scenarios where conflicts occur frequently.
```

Comment:
```suggestion
When we are performing a write operation, we try to lock every affected row in the `Prewrite` phase. If we failed to lock some of these rows, the whole transaction will be rolled back. As an optimistic lock algorithm, Percolator's transactional write algorithm may cause performance regressions in scenarios where conflicts occur frequently.
```

Reply:
Seems "cause" is not very accurate here. How about:
> ... As an optimistic lock algorithm, sometimes Percolator's transactional write algorithm may meet performance regressions in scenarios where conflicts occur frequently.

Reply:
What about "Using an optimistic lock algorithm, sometimes Percolator's transactional write may encounter performance regressions in scenarios where conflicts occur frequently."
This way we don't need to repeat "algorithm" twice.